### PR TITLE
kafka-4767: KafkaProducer is not joining its IO thread properly

### PR DESCRIPTION
**Pull request by huxihx**
KafkaProducer#close swallows the InterruptedException which might be acceptable when it's invoked from within the main thread or user is extending Thread and therefore controls all the code higher up on the call stack. For other cases, it'd better retstore the interupted status after capturing the exception.

### CODE REVIEW

**Comment by huxihx** [comment]
@ijuma Please have a review. Thanks.

**Comment by huxihx** [comment]
@ijuma Please have a reivew on this PR. Thanks.

**Comment by omkreddy** [comment]
LGTM

**Comment by ijuma** [comment]
Thanks for the PR. I looked this in more detail and it looks like we eventually throw `KafkaException` for this case. In the consumer, we throw `InterruptException` (which is a non-checked version of `InterruptedException` that inherits from `KafkaException`). Seems like we should do the same here. That class sets the interrupt in the constructor.

**Comment by huxihx** [comment]
@ijuma Followed the same pattern as what KafkaConsumer#close treats interruption but also explicitly added an if clause to check InterruptedException since `firstException` would be set to it explicitly in  KafkaProducer#close. Please have a review on that. Thanks.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 1]
```diff
@@ -730,8 +730,15 @@ private void close(long timeout, TimeUnit timeUnit, boolean swallowException) {
         ClientUtils.closeQuietly(partitioner, "producer partitioner", firstException);
         AppInfoParser.unregisterAppInfo(JMX_PREFIX, clientId);
         log.debug("The Kafka producer has closed.");
-        if (firstException.get() != null && !swallowException)
-            throw new KafkaException("Failed to close kafka producer", firstException.get());
+        Throwable exception = firstException.get();
+        if (exception != null && !swallowException) {
+            if (exception instanceof InterruptedException) {
+                throw new InterruptException("kafka producer thread was interrupted");
+            } else if (exception instanceof InterruptException) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Can this case ever happen?

Comment (huxihx) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: @ijuma , sorry for the delay. I am not sure actually. Just the defensive programming. Do you think it should be refined?

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: I'd just remove it I think. It's dead code and a bit confusing.

**Comment by huxihx** [comment]
@ijuma Well, already removed those dead code and also added the code to reserve the interruption status. Looks good now?

**Pull request closed without merging**